contracts/USM.sol

```diff
@@ -1,117 +1,136 @@
+        _setLatestFumPrice();
+        return usmMinted;
     }
 
     /**
-     * @notice Burn USM for ETH.
+     * @notice Burn USM for ETH with checks and asset transfers.
      *
-     * @param _usmAmount Amount of USM to burn.
+     * @param _usmToBurn Amount of USM to burn.
+     * @return ETH sent
      */
-    function burn(uint _usmAmount) external {
-        uint ethAmount = _usmToEth(_usmAmount);
-        uint ethMinusFee = ethAmount.sub(mulFixed(ethAmount, BURN_FEE));
-        ethPool = ethPool.sub(ethMinusFee);
-        _burn(msg.sender, _usmAmount);
-        Address.sendValue(msg.sender, ethMinusFee);
+    function burn(uint _usmToBurn) external returns (uint) {
+        require(_usmToBurn >= MIN_BURN_AMOUNT, "Must burn at least 1 USM");
+        uint ethToSend = _burn(_usmToBurn);
+        require(debtRatio() <= WAD,
+            "Cannot burn with debt ratio below 100%");
+        Address.sendValue(msg.sender, ethToSend);
+        // set latest fum price
+        _setLatestFumPrice();
+        return (ethToSend);
     }
 
     /**
-     * @notice Calculate debt ratio of the current Eth pool amount and outstanding USM
-     * (the amount of USM in total supply).
-     *
-     * @return Debt ratio in UNIT.
+     * @notice Funds the pool with ETH, minting FUM at its current price and considering if the debt ratio goes from under to over
      */
-    function debtRatio() external view returns (uint) {
-        if (ethPool == 0) {
-            return 0;
+    function fund() external payable {
+        require(msg.value > MIN_ETH_AMOUNT, "Must deposit more than 0.001 ETH");
+        if(debtRatio() > MAX_DEBT_RATIO){
+            uint ethNeeded = _usmToEth(totalSupply()).wadDiv(MAX_DEBT_RATIO).sub(ethPool).add(1); //+ 1 to tip it over the edge
+            console.log(ethNeeded);
+            if (msg.value >= ethNeeded) { // Split into two fundings at different prices
+                _fund(msg.sender, ethNeeded);
+                _fund(msg.sender, msg.value.sub(ethNeeded));
+                return;
+            } // Otherwise continue for funding the total at a single price
         }
-        // If divFixed is fed two integers, returns their division as a fixed point number
-        return divFixed(totalSupply(), _ethToUsm(ethPool));
+        console.log("debt ratio was below max");
+        _fund(msg.sender, msg.value);
     }
 
     /**
-     * @notice Convert ETH amount to USM using the latest price of USM
-     * in ETH.
-     *
-     * @param _ethAmount The amount of ETH to convert.
-     * @return The amount of USM.
+     * @notice Funds the pool with ETH, minting FUM at its current price
      */
-    function _ethToUsm(uint _ethAmount) internal view returns (uint) {
-        require(_ethAmount > 0, "Eth Amount must be more than 0");
-        return mulFixed(_oraclePrice(), _ethAmount);
+    function _fund(address to, uint ethIn) internal {
+        uint _fumPrice = fumPrice();
+        uint fumOut = _fumPrice.wadMul(ethIn);
```

Comment:
There's a bug here: should be `fumOut = ethIn.wadDiv(_fumPrice)`.  Eg, if the user passes in 3 ETH, and the current FUM price (in ETH terms) is 0.01 ETH, then the number of FUM created should be 3 / 0.01 = 300, not 0.01 * 3 = 0.03.